File: cmd/basic-vad-web/static/js/backends/pyannote.js
// Browser-side Pyannote Segmentation 3.0. Port of pkg/vad/pyannote.go.
//
// 10-second windows of raw float32 audio → log-softmax over 7 powerset
// classes per 17 ms frame (589 frames per window) → per-speaker segment
// extraction with cross-window merging.
//
// Powerset decoding maps argmax class index → set of active speakers per
// frame:
//   0: silence, 1: {A}, 2: {B}, 3: {A,B}, 4: {C}, 5: {A,C}, 6: {B,C}.
// (Max 2 of 3 speakers active per frame — pyannote 3.0 trained capacity.)

const SAMPLE_RATE = 16000;
const WINDOW_SIZE = 10 * SAMPLE_RATE; // 160_000 samples
const NUM_FRAMES = 589;
const NUM_CLASSES = 7;
const NUM_SPEAKERS = 3;

// Bit-masks: powersetMaskByClass[c] has bit `spk` set iff class c implies
// speaker `spk` is active. Faster than nested loops over the variable-length
// arrays in the Go code.
const POWERSET_MASK_BY_CLASS = new Uint8Array([
    0b000, // 0: silence
    0b001, // 1: A
    0b010, // 2: B
    0b011, // 3: A+B
    0b100, // 4: C
    0b101, // 5: A+C
    0b110, // 6: B+C
]);

export const PYANNOTE_AUX_FILES = []; // model.onnx only

export async function run(ort, session, samples) {
    if (!samples || samples.length === 0) return [];

    const allSegments = [];
    // Reuse an input buffer; ort.Tensor copies, so we can clobber it between
    // calls without invalidating earlier ones.
    const inputBuf = new Float32Array(WINDOW_SIZE);

    for (let offset = 0; offset < samples.length; offset += WINDOW_SIZE) {
        // Zero-fill, then copy whatever's available.
        inputBuf.fill(0);
        const copyEnd = Math.min(offset + WINDOW_SIZE, samples.length);
        inputBuf.set(samples.subarray(offset, copyEnd));

        const inputT = new ort.Tensor('float32', inputBuf, [1, 1, WINDOW_SIZE]);
        const out = await session.run({ input_values: inputT });
        const logits = out.logits.data;
        if (logits.length !== NUM_FRAMES * NUM_CLASSES) {
            throw new Error(`pyannote: logits len ${logits.length} != ${NUM_FRAMES * NUM_CLASSES}`);
        }
        const windowStart = offset / SAMPLE_RATE;
        const segs = extractSegments(logits, windowStart);
        for (const s of segs) allSegments.push(s);
    }
    return mergeSegments(allSegments);
}

function extractSegments(logits, windowStartSec) {
    const frameDuration = 10.0 / NUM_FRAMES;
    const segments = [];

    // Per-speaker active state. Use parallel arrays to avoid object thrash.
    const active = new Uint8Array(NUM_SPEAKERS);
    const starts = new Float64Array(NUM_SPEAKERS);
    const sumConf = new Float32Array(NUM_SPEAKERS);
    const counts = new Uint32Array(NUM_SPEAKERS);

    for (let frame = 0; frame < NUM_FRAMES; frame++) {
        const frameStart = windowStartSec + frame * frameDuration;
        const base = frame * NUM_CLASSES;

        // Argmax over the 7 log-probs.
        let maxIdx = 0;
        let maxLogP = logits[base];
        for (let c = 1; c < NUM_CLASSES; c++) {
            const v = logits[base + c];
            if (v > maxLogP) { maxLogP = v; maxIdx = c; }
        }
        const conf = Math.exp(maxLogP);
        const mask = POWERSET_MASK_BY_CLASS[maxIdx];

        for (let spk = 0; spk < NUM_SPEAKERS; spk++) {
            const isActive = (mask & (1 << spk)) !== 0;
            if (isActive) {
                if (!active[spk]) {
                    active[spk] = 1;
                    starts[spk] = frameStart;
                    sumConf[spk] = 0;
                    counts[spk] = 0;
                }
                sumConf[spk] += conf;
                counts[spk]++;
            } else if (active[spk]) {
                segments.push({
                    start: starts[spk],
                    end: frameStart,
                    speaker_id: spk,
                    confidence: sumConf[spk] / counts[spk],
                });
                active[spk] = 0;
            }
        }
    }

    const endTime = windowStartSec + 10.0;
    for (let spk = 0; spk < NUM_SPEAKERS; spk++) {
        if (active[spk]) {
            segments.push({
                start: starts[spk],
                end: endTime,
                speaker_id: spk,
                confidence: sumConf[spk] / counts[spk],
            });
        }
    }
    return segments;
}

// mergeSegments merges adjacent same-speaker segments across window boundaries
// when the gap is < 0.1 s. Mirrors mergePyannoteSegments in Go.
function mergeSegments(segments) {
    if (segments.length === 0) return [];
    const out = [];
    for (const seg of segments) {
        if (out.length > 0) {
            const last = out[out.length - 1];
            if (last.speaker_id === seg.speaker_id && seg.start - last.end < 0.1) {
                last.end = seg.end;
                continue;
            }
        }
        out.push({ ...seg });
    }
    return out;
}

File: pkg/vad/pyannote.go
package vad

import (
	"fmt"
	"math"
	"sync"

	ort "github.com/yalue/onnxruntime_go"
)

// pyannote-specific powerset class mapping for segmentation 3.0:
// 3 speakers, max 2 active per frame → 7 classes.
//
//	class 0: ∅ (silence)
//	class 1: {A}
//	class 2: {B}
//	class 3: {A, B}
//	class 4: {C}
//	class 5: {A, C}
//	class 6: {B, C}
var powersetMapping = [pyannoteNumClasses][]int{
	0: {},     // silence
	1: {0},    // A
	2: {1},    // B
	3: {0, 1}, // A+B
	4: {2},    // C
	5: {0, 2}, // A+C
	6: {1, 2}, // B+C
}

const (
	pyannoteNumSpeakers = 3
	// WindowSize is 10 seconds of audio at 16 kHz — the pyannote-3.0 window length.
	WindowSize = 10 * SampleRate // 160000 samples
	// pyannoteNumFrames is the output frame count for one WindowSize input.
	pyannoteNumFrames = 589
	pyannoteNumClasses = 7
)

// Model is the Pyannote Segmentation 3.0 backend. Implements Backend.
//
// Kept as the type name `Model` (instead of `Pyannote`) for backward compat
// with the original vad/server.go API; cmd/vad's default backend constructs
// this via NewModel.
type Model struct {
	mu      sync.Mutex
	session *ort.AdvancedSession
	input   *ort.Tensor[float32]
	output  *ort.Tensor[float32]
}

// NewModel creates a new Pyannote VAD model from an ONNX file.
//
// Equivalent to NewPyannote — kept under the original name so older callers
// keep compiling without modification.
func NewModel(modelPath string) (*Model, error) {
	inputShape := ort.NewShape(1, 1, int64(WindowSize))
	outputShape := ort.NewShape(1, int64(pyannoteNumFrames), int64(pyannoteNumClasses))

	input, err := ort.NewEmptyTensor[float32](inputShape)
	if err != nil {
		return nil, fmt.Errorf("create input tensor: %w", err)
	}

	output, err := ort.NewEmptyTensor[float32](outputShape)
	if err != nil {
		input.Destroy()
		return nil, fmt.Errorf("create output tensor: %w", err)
	}

	session, err := ort.NewAdvancedSession(modelPath,
		[]string{"input_values"}, []string{"logits"},
		[]ort.Value{input}, []ort.Value{output},
		nil,
	)
	if err != nil {
		input.Destroy()
		output.Destroy()
		return nil, fmt.Errorf("create session: %w", err)
	}

	return &Model{session: session, input: input, output: output}, nil
}

// NewPyannote is an alias for NewModel for clarity at the call site once
// multiple backends exist.
func NewPyannote(modelPath string) (*Model, error) { return NewModel(modelPath) }

// Close releases all resources.
func (m *Model) Close() {
	m.mu.Lock()
	defer m.mu.Unlock()
	if m.session != nil {
		m.session.Destroy()
	}
	if m.input != nil {
		m.input.Destroy()
	}
	if m.output != nil {
		m.output.Destroy()
	}
}

// ProcessAudio runs VAD on raw 16 kHz mono float32 audio samples. It processes
// the audio in 10-second windows and returns detected segments.
func (m *Model) ProcessAudio(samples []float32) ([]Segment, error) {
	m.mu.Lock()
	defer m.mu.Unlock()

	var allSegments []Segment

	for offset := 0; offset < len(samples); offset += WindowSize {
		inputData := m.input.GetData()
		for i := range inputData {
			inputData[i] = 0
		}

		copyEnd := offset + WindowSize
		if copyEnd > len(samples) {
			copyEnd = len(samples)
		}
		copy(inputData, samples[offset:copyEnd])

		if err := m.session.Run(); err != nil {
			return nil, fmt.Errorf("inference at offset %d: %w", offset, err)
		}

		windowStart := float64(offset) / float64(SampleRate)
		segments := extractPyannoteSegments(m.output.GetData(), windowStart)
		allSegments = append(allSegments, segments...)
	}

	return mergePyannoteSegments(allSegments), nil
}

// extractPyannoteSegments converts log-softmax output into per-speaker voice
// segments using powerset decoding.
func extractPyannoteSegments(logits []float32, windowStartSec float64) []Segment {
	frameDuration := 10.0 / float64(pyannoteNumFrames)

	var segments []Segment

	type speakerState struct {
		active  bool
		start   float64
		sumConf float32
		count   int
	}
	states := make([]speakerState, pyannoteNumSpeakers)

	for frame := 0; frame < pyannoteNumFrames; frame++ {
		frameStart := windowStartSec + float64(frame)*frameDuration

		maxIdx := 0
		maxLogP := logits[frame*pyannoteNumClasses]
		for c := 1; c < pyannoteNumClasses; c++ {
			if logits[frame*pyannoteNumClasses+c] > maxLogP {
				maxLogP = logits[frame*pyannoteNumClasses+c]
				maxIdx = c
			}
		}
		confidence := float32(math.Exp(float64(maxLogP)))
		activeSpeakers := powersetMapping[maxIdx]

		for spk := 0; spk < pyannoteNumSpeakers; spk++ {
			isActive := false
			for _, s := range activeSpeakers {
				if s == spk {
					isActive = true
					break
				}
			}

			if isActive {
				if !states[spk].active {
					states[spk].active = true
					states[spk].start = frameStart
					states[spk].sumConf = 0
					states[spk].count = 0
				}
				states[spk].sumConf += confidence
				states[spk].count++
			} else if states[spk].active {
				segments = append(segments, Segment{
					Start:      states[spk].start,
					End:        frameStart,
					SpeakerID:  spk,
					Confidence: states[spk].sumConf / float32(states[spk].count),
				})
				states[spk].active = false
			}
		}
	}

	endTime := windowStartSec + 10.0
	for spk := 0; spk < pyannoteNumSpeakers; spk++ {
		if states[spk].active {
			segments = append(segments, Segment{
				Start:      states[spk].start,
				End:        endTime,
				SpeakerID:  spk,
				Confidence: states[spk].sumConf / float32(states[spk].count),
			})
		}
	}

	return segments
}

// mergePyannoteSegments merges adjacent segments for the same speaker across
// window boundaries.
func mergePyannoteSegments(segments []Segment) []Segment {
	if len(segments) == 0 {
		return nil
	}
	var merged []Segment
	for _, seg := range segments {
		if len(merged) > 0 {
			last := &merged[len(merged)-1]
			if last.SpeakerID == seg.SpeakerID && seg.Start-last.End < 0.1 {
				last.End = seg.End
				continue
			}
		}
		merged = append(merged, seg)
	}
	return merged
}

// compile-time check that *Model satisfies Backend.
var _ Backend = (*Model)(nil)
